### PR TITLE
ignoring namespaces by matching regex

### LINKED ISSUE #161: Disable namespaces by regex
Allow for `ignoreNamespaces` [configurable](https://github.com/spiffe/spire-controller-manager/blob/main/docs/spire-controller-manager-config.md) to take a regex for namespaces to disable.

Ported over from https://github.com/spiffe/spire/issues/3386

### PR DESCRIPTION
**Pull request by unnathik**
- Created a new function to match regex with strings in a stringset
- Replaced uses of `In` function to match regex instead

fixes #161 and #175

### CODE REVIEW

**Comment by faisal-memon** [comment]
@unnathik Thanks for submitting this PR.

File: pkg/stringset/stringset.go
Comment (faisal-memon) [inline pkg/stringset/stringset.go:36]: We need to handle the error in case the regex is malformed. I think the best way is to compile the regex at boot time using [regex.Compile()](https://pkg.go.dev/regexp#Compile). We can do so here in main go here after reading in the config file: https://github.com/spiffe/spire-controller-manager/blob/2b0ff48b84150cb1451234af111a0a2676134b3d/main.go#L104-L116

Step through ctrlConfig.IgnoreNamespaces and create a `[]regexp.Regex` and then pass this to the podReconciler instead of the array of strings:
https://github.com/spiffe/spire-controller-manager/blob/2b0ff48b84150cb1451234af111a0a2676134b3d/main.go#L291-L299

Then in `MatchRegex()` you can use [FindString()](https://pkg.go.dev/regexp#Regexp.FindString) which doesn't return err.

Comment (unnathik) [reply pkg/stringset/stringset.go:36]: If the regex is malformed, do we want to set ignoreNamespaces to the default value or only throw an error?

Comment (faisal-memon) [reply pkg/stringset/stringset.go:36]: Throw error

**Comment by unnathik** [comment]
Made the discussed changes; also wasn't sure if the regex matching should become an independent function since it's used twice but can create the function if required

File: controllers/pod_controller.go
Comment (faisal-memon) [inline controllers/pod_controller.go:52]:
```suggestion
		if regex.FindString(req.Namespace) == "" {
```

Comment (faisal-memon) [inline controllers/pod_controller.go:52]:
```suggestion
		if regex.FindString(req.Namespace) != "" {
```

File: main.go
Comment (faisal-memon) [inline main.go:66]: I think it will be better to make this `[]*regexp.Regexp`. I didn't realize `Compile()` returned a pointer.

File: main.go
Comment (faisal-memon) [inline main.go:122]:
```suggestion
				return ctrlConfig, options, fmt.Errorf("unable to compile ignore namespaces regex: %w", err)
```

File: pkg/spireentry/reconciler.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler.go:127]: I dont think you need this extra `ignored` value. You can `continue` after the first regex match, that's what was being done before.

File: pkg/stringset/stringset.go
Comment (faisal-memon) [inline pkg/stringset/stringset.go:19]: I think this file is not longer needed?

File: main.go
Comment (faisal-memon) [inline main.go:119]: Can you use something more descriptive the s?

File: main.go
Comment (faisal-memon) [reply main.go:66]: Also can you move this out of this global var section and on top of the for loop where it is populated?

Comment (unnathik) [reply main.go:66]: > Also can you move this out of this global var section and on top of the for loop where it is populated?

Since it is being referred outside of the function with the for loop where it is being populated, wouldn't it become inaccessible to the other functions if it was moved out of the global var section?

File: pkg/spireentry/reconciler.go
Comment (unnathik) [reply pkg/spireentry/reconciler.go:127]: Since there are two for loops now, continue would move onto the next iteration of the inner loop; I added the ignored value to move onto the next iteration of the outer loop (continue) after the first regex match.

File: pkg/stringset/stringset.go
Comment (unnathik) [reply pkg/stringset/stringset.go:19]: Yes, removing it now.

File: pkg/spireentry/reconciler.go
Comment (unnathik) [reply pkg/spireentry/reconciler.go:127]: > if i'm following it fine, once you get a single one that must be ignored all next namapesces will be ignored too, is that right? and my we use regex.MathString instead?

That's correct--once the namespace matches the regex pattern for a namespace it should ignore, it won't check if it matches the pattern for other ignored namespaces. 

We used regex.MatchString earlier (commit `e330c4d381f8899730d2dea6dcab47b9eed8c35a`), but Faisal said there might be issues with error handling there. We discussed it here: https://github.com/spiffe/spire-controller-manager/pull/170#discussion_r1235720412.

File: main.go
Comment (faisal-memon) [reply main.go]: Yes, which is fine. You can pass it around as an parameter.

File: pkg/spireentry/reconciler.go
Comment (faisal-memon) [reply pkg/spireentry/reconciler.go:127]: @MarcosDY I think its better to compile the regex at boot so if there's a bad regex we can error out right away.

Comment (faisal-memon) [reply pkg/spireentry/reconciler.go:127]: @unnathik Good point. I think we can move the inner for loop to a function. I think it will be easier to read that way.

**Comment by faisal-memon** [comment]
@MarcosDY Can you enable the workflow run on this?

File: pkg/spireentry/reconciler.go
Comment (unnathik) [reply pkg/spireentry/reconciler.go]: @MarcosDY good catch. Made changes to address both comments in latest commit.

File: controllers/pod_controller.go
Comment (azdagron) [inline controllers/pod_controller.go:52]: I think what @MarcosDY was saying is that this call here can use the MatchString method on the compiled regex, which does not return an error. FindString is also fine just less intuitive since the result is throwaway.
```suggestion
		if regex.MatchString(req.Namespace) {
```

File: pkg/spireentry/reconciler.go
Comment (azdagron) [inline pkg/spireentry/reconciler.go:218]:
```suggestion
		if regex.MatchString(namespace) {
```

File: pkg/spireentry/reconciler.go
Comment (azdagron) [inline pkg/spireentry/reconciler.go:113]: This variable is not used outside of the scope of the for loop and should ideally be declared close to the point of use.

File: pkg/spireentry/reconciler.go
Comment (azdagron) [inline pkg/spireentry/reconciler.go:118]: The ignored variable is used in a single spot. I'm curious if the code isn't more clear if we just use the function call directly in the if statement. It might help readability if we rename the function slightly:
```suggestion
			if isNamespaceIgnored(r.config.IgnoreNamespaces, namespaces[i].Name) {
```

**Review by azdagron — APPROVED** [review]
I've approved but left some nitpicky comments. No pressure to make the changes if you don't have the inclination. Thanks for doing this @unnathik !

**Comment by azdagron** [comment]
Before we can merge, we need DCO signoff on all of the commits. Is that something you can fix up @unnathik ?

File: pkg/spireentry/reconciler.go
Comment (azdagron) [inline pkg/spireentry/reconciler.go:215]: This is now broken. isNamespaceIgnored should return true if the regex is a match.

Comment (unnathik) [reply pkg/spireentry/reconciler.go]: didn't realize the code broke--sorry! fixed now.

File: pkg/spireentry/reconciler.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler.go:213]: Sorry i missed this before, can you move this function to the end of the file? Its not part of the struct.

**Comment by unnathik** [comment]
Added the function to the end and created some tests for the `isNamespaceIgnored` function for #175

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:85]: Can we make this an anonymous struct?

```
tests := []struct {
   ignoredNamespaces []string
   namespace string
   expected bool
} {
  // test case 1
  // test case 2
}

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:89]: Rather than compile these here. let's put the uncompiled string in the test case, then compile in the for loop.

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:102]:
```suggestion
                require.Equal(t, test.expected, actual)
```

Comment (faisal-memon) [reply pkg/spireentry/reconciler_test.go:102]: Can also add a message. see https://pkg.go.dev/github.com/stretchr/testify/require#Equal and https://pkg.go.dev/github.com/stretchr/testify/require#Equalf

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:99]:
```suggestion
			require.NoError(t, err)
			regexIgnoredNamespaces = append(regexIgnoredNamespaces, regex)
```

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:89]: It would be good to vary the regex strings

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:104]: I think when equalf fails it outputs expected and actual.

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [reply pkg/spireentry/reconciler_test.go:89]: nm the above. Lets keep the regex strings constant and vary just `namespace` and `expected`. Can you move the regex string above the test struct and compile it. Then go through the test cases. That should work better.

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:93]:
```suggestion
	ignoredNamespaces := []*regexp.Regexp{
		regexp.MustCompile("s([a-z]+)re"),
		regexp.MustCompile("default"),
	}
	
	tests := []struct {
		namespace         string
		expected          bool
	}{
		{"spire", true},
		{"default", true},
		{"spiffe", false},
		{"kubernetes", false},
	}
```

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [inline pkg/spireentry/reconciler_test.go:96]:
```suggestion
		actual := isNamespaceIgnored(ignoredNamespaces, test.namespace)
```

File: pkg/spireentry/reconciler_test.go
Comment (faisal-memon) [reply pkg/spireentry/reconciler_test.go:93]: Thanks for making the changes. Small change here to reduce duplication a bit.

**Review by faisal-memon — APPROVED** [review]
Thanks @unnathik for adding the unit tests. This looks good to me.

File: main.go
Comment (MarcosDY) [inline main.go:125]: nit: instead of creating a global var, what do you think about returning this regex list on parseConfig? and pass it to required functions?

File: controllers/pod_controller.go
Comment (MarcosDY) [inline controllers/pod_controller.go:51]: nit: it looks like `isNamespaceIgnored` can be moved o a common package and used here too

**Review by MarcosDY — APPROVED** [review]

**Pull request merged**
